src/GMCPAuthenticator.cpp

```diff
@@ -0,0 +1,110 @@
+ *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
+ ***************************************************************************/
+
+#include "GMCPAuthenticator.h"
+
+#include "Host.h"
+#include "ctelnet.h"
+#include <QDebug>
+
+GMCPAuthenticator::GMCPAuthenticator(Host* pHost) : mpHost(pHost) {}
+
+void GMCPAuthenticator::saveSupportsSet(const QString& data)
+{
+    auto jsonDoc = QJsonDocument::fromJson(data.toUtf8());
+    auto jsonObj = jsonDoc.object();
+
+    if (jsonObj.contains("type")) {
+        QJsonArray typesArray = jsonObj["type"].toArray();
+        for (const auto& type : typesArray) {
+            mSupportedAuthTypes.append(type.toString());
+        }
+    }
+
+    qDebug() << "Supported auth types:" << mSupportedAuthTypes;
+}
+
+void GMCPAuthenticator::sendCredentials()
+{
+    auto character = mpHost->getLogin();
+    auto password = mpHost->getPass();
+    QJsonObject credentials;
+    if (!character.isEmpty() && !password.isEmpty()) {
+        credentials["account"] = character;
+        credentials["password"] = password;
+    }
+    QJsonDocument doc(credentials);
+    QString gmcpMessage = doc.toJson(QJsonDocument::Compact);
+
+    std::string output;
+    output += TN_IAC;
+    output += TN_SB;
+    output += OPT_GMCP;
+    output += "Char.Login.Credentials ";
+    output += mpHost->mTelnet.encodeAndCookBytes(gmcpMessage.toStdString());
+    output += TN_IAC;
+    output += TN_SE;
+
+    // Send credentials to server
+    mpHost->mTelnet.socketOutRaw(output);
+    qDebug() << "Sent GMCP credentials";
+}
+
+
+void GMCPAuthenticator::handleAuthResult(const QString& data)
+{
+    auto doc = QJsonDocument::fromJson(data.toUtf8());
+    auto obj = doc.object();
+
+    bool success = obj["success"].toBool();
+    auto message = obj["message"].toString();
+
+    if (success) {
+        qDebug() << "GMCP login successful";
+    } else {
+        qDebug() << "GMCP login failed:" << message;
+        if (message.isEmpty()) {
+            mpHost->postMessage(tr("[ WARN ]  - Could not log in to the game, is the login information correct?"));
+        } else {
+            mpHost->postMessage(tr("[ WARN ]  - Could not log in to the game: %1").arg(message));
+        }
+
+    }
+}
+
+// controller for GMCP authentication
+void GMCPAuthenticator::handleAuthGMCP(const QString& packageMessage, const QString& data)
+{
+    if (packageMessage == qsl("Char.Login.Default")) {
+        saveSupportsSet(data);
+
+        if (mSupportedAuthTypes.contains(qsl("password-credentials"))) {
+            mpHost->mTelnet.cancelLoginTimers();
+            sendCredentials();
+        } else {
+            qDebug() << "Server does not support credentials authentication and we don't support any other";
+        }
+
+    } else if (packageMessage == qsl("Char.Login.Result")) {
+        handleAuthResult(data);
+
+    } else {
+        qDebug() << "Unknown GMCP auth package:" << packageMessage;
+    }
```

Comment:
Invoking the no "else-after-return" paradigm: 
```suggestion
        return;
    }

    if (packageMessage == qsl("Char.Login.Result")) {
        handleAuthResult(data);
        return;
    }
    
    qDebug() << "Unknown GMCP auth package:" << packageMessage;
```

Reply:
I am not sure we have to force the paradigm. Rather, if you have an if with a return, don't add an else.

Reply:
It is just that nested `} else if () {` can be confusing if done too deeply - and this method can be refactored in the manner I suggested which does make it easier to parse - yeah I know that doesn't sound like *me* but I guess I have seen the light at the end of the tunnel (though it might be a train coming in the opposite direction...) :wink: